tracer/src/Datadog.Trace/Configuration/TracerSettingsSnapshot.cs

```diff
@@ -0,0 +1,328 @@
+            var oldValue = Integrations.Settings[i];
+            var integrationName = newValue.IntegrationNameInternal;
+            RecordIfChanged(telemetry, string.Format(ConfigurationKeys.Integrations.Enabled, integrationName), oldValue.EnabledInternal, newValue.EnabledInternal);
+#pragma warning disable 618 // App analytics is deprecated, but still used
+            RecordIfChanged(telemetry, string.Format(ConfigurationKeys.Integrations.AnalyticsEnabled, integrationName), oldValue.AnalyticsEnabledInternal, newValue.AnalyticsEnabledInternal);
+            RecordIfChanged(telemetry, string.Format(ConfigurationKeys.Integrations.AnalyticsSampleRate, integrationName), oldValue.AnalyticsSampleRateInternal, newValue.AnalyticsSampleRateInternal);
+#pragma warning restore 618
+        }
+
+        Exporter.RecordChanges(settings.ExporterInternal, telemetry);
+    }
+
+    private static void RecordIfChanged(IConfigurationTelemetry telemetry, string key, string? oldValue, string? newValue)
+    {
+        if (!string.Equals(oldValue, newValue, StringComparison.Ordinal))
+        {
+            telemetry.Record(key, newValue, recordValue: true, ConfigurationOrigins.Code);
+        }
+    }
+
+    private static void RecordIfChanged(IConfigurationTelemetry telemetry, string key, int? oldValue, int? newValue)
+    {
+        if (oldValue != newValue)
+        {
+            telemetry.Record(key, newValue, ConfigurationOrigins.Code);
+        }
+    }
+
+    private static void RecordIfChanged(IConfigurationTelemetry telemetry, string key, double? oldValue, double? newValue)
+    {
+        if (oldValue != newValue)
+        {
+            telemetry.Record(key, newValue, ConfigurationOrigins.Code);
+        }
+    }
+
+    private static void RecordIfChanged(IConfigurationTelemetry telemetry, string key, bool oldValue, bool newValue)
+    {
+        if (oldValue != newValue)
+        {
+            telemetry.Record(key, newValue, ConfigurationOrigins.Code);
+        }
+    }
+
+    private static void RecordIfChanged(IConfigurationTelemetry telemetry, string key, bool? oldValue, bool? newValue)
+    {
+        if (oldValue != newValue)
+        {
+            if (newValue is null)
+            {
+                telemetry.Record(key, null, recordValue: true, ConfigurationOrigins.Code);
+            }
+            else
+            {
+                telemetry.Record(key, newValue.Value, ConfigurationOrigins.Code);
+            }
+        }
+    }
+
+    private static void RecordIfChanged(IConfigurationTelemetry telemetry, string key, IDictionary<string, string>? oldValues, IDictionary<string, string>? newValues)
+    {
+        if (oldValues is null
+         && (newValues is null || newValues is { Count: 0 }))
+        {
+            return;
+        }
+
+        var equal = true;
+
+        if (oldValues is not null
+         && newValues is not null
+         && oldValues.Count == newValues.Count)
```

Comment:
Shouldn't equal be false if the dictionaries have a different number of values?

(looks like you implemented it correctly for Hashset)

Reply:
oops, that was me trying to work out why it lies about the nullability in the if clause, thanks!